### DISCUSSION TITLE
カップラーメンタイマ(base1)のタイマ4の設定について

### DISCUSSION THREAD

**Discussion by Shinya-Omoto** [toppers/users-forum#125] [TOPPERS Seminar]
お世話になっております。
Toppersセミナーを受講しております、小元と申します。

base1\program\cup_timerのタイマ4の設定について、
不明点があり質問させていただきました。

現在、base1のカップラーメンタイマを使い、マイコン制御の学習を行っているのですが、
タイマ4で10ms毎に割り込みを行うように変更しようと思い、以下のような設定を行いました。

PrescalerValue = (84,000,000 / 4000) - 1 　// 20,999
*TREG_TIM4_ARR= 40 - 1

しかし、オシロスコープを使用して時間計測を行ったところ、9.875msでの割り込みとなってしまいました。
上記のように設定しましたが、何か間違いなどありますでしょうか？
間違いがあればご教示頂けますと幸いです。

よろしくお願いいたします。

**Comment by Ryohsuke-Takeuchi** [comment]
申し訳ありません。リファレンスマニュアルを見ると
TREG_TIM4_ARRの設定値は、-1しない値を設定するようです。

上記の設定では
PrescalerValue の値は、21000であり
リロード設定が39なので
21000/84,000,000 * 39 = 0.00975で9.75msとなります。（CDK=0B00の場合）

PrescalerValue が精度なので、この値を小さくして、リロード設定を大きくすると
精度が高くなります。（但し、この設定は16bitのレジスタに正しく設定されなければなりません）
PrescalerValue の値は、21000であれば、0.25ms単位のカウントアップで精度上で、
0.25msは10msで割り切れる値なので、範囲の誤差はシステムクロック84Mhzの精度による微小な誤差以外
は発生しないと思います。

int_timer.cを使って、上記の設定でLED02のGPIOを測定したところ
9.854msとなりました。

![10ms](https://github.com/toppers/users-forum/assets/91311953/be1d6ff2-313e-40aa-bcb6-f6408c0af6f1)

*TREG_TIM4_ARR= 40 とした場合、
21000/84,000,000 * 40 = 0.010で10msとなります。（CDK=0B00の場合）
実測すると、以下のように10.0066msとなりました。
TREG_TIM4_ARRの設定値39,40の場合でも設定時間より少し大きいのは、タイマー割込みで割込み受けつけ後、
	*TREG_TIM4_SR = 0x0000;
で割込みクリアして、ここでリロードが発生するようです。
いずれにしても、精度をもとめるならば、割込みハンドラ実行時間を考慮したタイマー時間を設定する必要があると思います。

![10-2](https://github.com/toppers/users-forum/assets/91311953/1e59ee63-2b5b-4f93-b34f-009586304a0b)

以下のように、割込みクリアの前に、約100μsecのDelayをいれると
	sil_dly_nse(100*1000);
	*TREG_TIM4_SR = 0x0000;
実測値は、10.119msとなりました。

![10-3](https://github.com/toppers/users-forum/assets/91311953/ead6df99-f2e9-488f-8187-3beafe567cba)
